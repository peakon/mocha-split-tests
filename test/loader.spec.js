const {
  getTestFilesInfo,
  getFileRuntimeStats,
  attachRuntimesToFiles
} = require("../lib/loader");

describe("loader", function() {
  describe("getTestFilesInfo", function() {
    it("should fail when no files are found", function() {
      const pattern = "**/doesnotexist/*.js";
      expect(() => getTestFilesInfo(pattern)).toThrow(
        `No files matching a pattern: ${pattern}`
      );
    });

    it("should return files when pattern is correct", function() {
      expect(getTestFilesInfo("test/data/**/*.spec")).toMatchSnapshot();
    });
  });

  describe("getFileRuntimeStats", function() {
    describe("when runtime.log contains malformed rows", function() {
      it("loads only well-formed rows from file", function() {
        const stats = getFileRuntimeStats("./test/data/runtimes2.log");
        expect(stats.size).toEqual(3);
        expect(stats).toMatchSnapshot();
      });
    });

    describe("when runtime.log has no malformed rows", function() {
      it("loads all stats from a well-formed file", function() {
        const stats = getFileRuntimeStats("./test/data/runtimes1.log");
        expect(stats.size).toEqual(11);
        expect(stats).toMatchSnapshot();
      });
    });

    describe("when runtime.log does not exist", function() {
      it("should not fail with error", function() {
        const stats = getFileRuntimeStats("/invalid/path/to/runtime.log");
        expect(stats.size).toEqual(0);
      });
    });
  });

  describe("attachRuntimesToFiles", function() {
    it("adds runtime stats to test backlog", function() {
      const testFiles = [
        { path: "test/1.spec.js", size: 1 },
        { path: "test/2.spec.js", size: 2 },
        { path: "test/3.spec.js", size: 3 }
      ];
      const runtimeStats = new Map()
        .set("test/1.spec.js", 100)
        .set("test/2.spec.js", 200)
        .set("test/4.spec.js", 400);
      expect(attachRuntimesToFiles(testFiles, runtimeStats)).toEqual([
        { path: "test/1.spec.js", size: 1, runtime: 100 },
        { path: "test/2.spec.js", size: 2, runtime: 200 },
        { path: "test/3.spec.js", size: 3, runtime: 0 }
      ]);
    });
  });
});
