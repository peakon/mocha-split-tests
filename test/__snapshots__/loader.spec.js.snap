// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader getFileRuntimeStats when runtime.log contains malformed rows loads only well-formed rows from file 1`] = `
Map {
  "test/controllers/application_controller.spec.js" => 5084,
  "test/models/user.spec.js" => 5084,
  "test/does-not-exist-anymore.spec.js" => 234,
}
`;

exports[`loader getFileRuntimeStats when runtime.log has no malformed rows loads all stats from a well-formed file 1`] = `
Map {
  "test/controllers/application_controller.spec.js" => 5084,
  "test/models/user.spec.js" => 5084,
  "test/models/company.spec.js" => 1234,
  "test/routers/foo.spec.js" => 20,
  "test/config/config.spec.js" => 345,
  "test/models/account.spec.js" => 2435,
  "test/routers/bar.spec.js" => 20,
  "test/config/buz.spec.js" => 345,
  "test/models/aaa.spec.js" => 2435,
  "test/does-not-exist-anymore.spec.js" => 234,
  "test/does-not-exist-anymore_2.spec.js" => 234234,
}
`;

exports[`loader getTestFilesInfo should return files when pattern is correct 1`] = `
Array [
  Object {
    "path": "test/data/fakeSpecs/2.spec",
    "size": 2,
  },
  Object {
    "path": "test/data/fakeSpecs/3.spec",
    "size": 3,
  },
  Object {
    "path": "test/data/fakeSpecs/4.spec",
    "size": 4,
  },
  Object {
    "path": "test/data/fakeSpecs/specs/1.spec",
    "size": 1,
  },
]
`;
